cache/nydus.go

```diff
@@ -0,0 +1,163 @@
+package cache
+
+import (
+	"context"
+	"os"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/mount"
+	digest "github.com/opencontainers/go-digest"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+
+	"github.com/moby/buildkit/session"
+	"github.com/moby/buildkit/util/compression"
+
+	"github.com/goharbor/acceleration-service/pkg/driver/nydus/backend"
+	"github.com/goharbor/acceleration-service/pkg/driver/nydus/packer"
+)
+
+const keyCachedNydusBootstrap = "cache.nydus-bootstrap"
+const keyCachedNydusBlob = "cache.nydus-blob"
+
+const valueCachedEmptySha256 = "sha256:none"
+
+type buildLayer struct {
+	sr *immutableRef
+	sg session.Group
+}
+
+func isEmptyNydusBlob(sr *immutableRef) bool {
+	return sr.GetString(keyCachedNydusBlob) == valueCachedEmptySha256
+}
+
+func (layer *buildLayer) getArtifact(ctx context.Context, dgst string, compressionType compression.Type) (*ocispecs.Descriptor, error) {
+	artifactDigest := digest.Digest(dgst)
+	if artifactDigest.Validate() != nil {
+		return nil, errdefs.ErrNotFound
+	}
+
+	artifactDesc, err := layer.sr.getBlobWithCompression(ctx, compressionType)
+	if err != nil {
+		return nil, errdefs.ErrNotFound
+	}
+
+	return &artifactDesc, nil
+}
+
+// GetCache get nydus bootstrap/blob descriptor from cache,
+// following situations should be returned:
+// err != nil, cache miss;
+// err == nil, cache hits;
+func (layer *buildLayer) GetCache(ctx context.Context) (*ocispecs.Descriptor, []ocispecs.Descriptor, error) {
+	bootstrapDesc, err := layer.getArtifact(ctx, layer.sr.GetString(keyCachedNydusBootstrap), compression.NydusBootstrap)
```

Comment:
Just want to double-check my understanding here. Each layer gets associated with a "bootstrap blob", but then in the final exported image there's only one bootstrap blob, which is merged from all the bootstrap blobs of the layers of the image?

Reply:
Correct!

- Nydus **blob** layer just represents the addition or modification file data of upper layer, no deletion data.
- Nydus **bootstrap** layer represents the whole metadata of filesystem view for entire image.